ProcessTree.cpp

```diff
@@ -0,0 +1,530 @@
+{
+    std::unique_lock<std::mutex> queue_pop_lock(_queue_pop_mutex);
+    while (!IsStopping()) {
+        _queue_data.wait(queue_pop_lock, [&]{return _queue_data_ready;});
+        while (!_PnQueue.empty()) {
+            struct ProcessQueueItem p = _PnQueue.front();
+            _PnQueue.pop();
+            switch (p.type) {
+                case ProcessQueueFork:
+                    AddPid(p.pid, p.ppid);
+                    break;
+                case ProcessQueueExec:
+                    AddPid(p.pid);
+                    break;
+                case ProcessQueueExit:
+                    RemovePid(p.pid);
+                    break;
+                default:
+                    Logger::Error("Invalid ProcessQueueType");
+            }
+        }
+
+        // Check if it's time for routine pruning of stale pids
+        std::chrono::duration<double> elapsed_seconds = std::chrono::system_clock::now() - _last_clean_time;
+        if (elapsed_seconds.count() > CLEAN_PROCESS_INTERVAL) {
+            Clean();
+            _last_clean_time = std::chrono::system_clock::now();
+        }
+        _queue_data_ready = false;
+    }
+}
+
+/* Process event from pnotify (fork)
+   If the pid is not in our table then add it.
+   If the ppid is in our table, copy the data from that entry.
+*/
+void ProcessTree::AddPid(int pid, int ppid)
+{
+    std::unique_lock<std::mutex> process_write_lock(_process_write_mutex);
+    if (_processes.count(pid) == 0) {
+        std::shared_ptr<ProcessTreeItem> process = std::make_shared<ProcessTreeItem>(ProcessTreeSource_pnotify, pid, ppid);
+        if (ppid && _processes.count(ppid) > 0) {
+            process->_uid = _processes[ppid]->_uid;
+            process->_gid = _processes[ppid]->_gid;
+            process->_exe = _processes[ppid]->_exe;
+            process->_cmdline = _processes[ppid]->_cmdline;
+            process->_exec_propagation = _processes[ppid]->_exec_propagation;
+            _processes[ppid]->_children.emplace_back(pid);
+            process->_ancestors = _processes[ppid]->_ancestors;
+            struct Ancestor anc = {ppid, _processes[ppid]->_exe};
+            process->_ancestors.emplace_back(anc);
+            ApplyFlags(process);
+        } else {
+            struct Ancestor anc = {ppid, ""};
+            process->_ancestors.emplace_back(anc);
+        }
+        _processes[pid] = process;
+    } 
+}
+
+/* Process event from pnotify (exec)
+   If process already exists (as it should) then increase the exec propagation variable in order to
+   make sure the next execve audit event fills in this processes' details.
+   If the process doesn't exist then add it.
+*/
+void ProcessTree::AddPid(int pid)
+{
+    std::unique_lock<std::mutex> process_write_lock(_process_write_mutex);
+    if (_processes.count(pid) > 0) {
+        if (_processes[pid]->_source == ProcessTreeSource_pnotify) {
+            _processes[pid]->_exec_propagation = _processes[pid]->_exec_propagation + 1;
```

Comment:
Why not "_processes[pid]->_exec_propagation += 1" ??

Reply:
Yep.